File: pol/api/v0/models/subject.py
import datetime
from typing import Dict, Optional

from pydantic import Field, BaseModel

from pol.api.v0.models.wiki import Wiki


class Rating(BaseModel):
    rank: int
    total: int
    count: Dict[str, int]
    score: float


class Images(BaseModel):
    large: str
    common: str
    medium: str
    small: str
    grid: str


class Collection(BaseModel):
    wish: int
    collect: int
    doing: int
    on_hold: int
    dropped: int


class Subject(BaseModel):
    id: int
    type: int
    name: str
    name_cn: str
    summary: str
    nsfw: bool
    locked: bool
    date: Optional[datetime.date]
    platform: str = Field(description="TV, Web, 欧美剧, PS4...")
    images: Optional[Images]
    infobox: Optional[Wiki]

    volumes: int = Field(description="书籍条目的册数，由旧服务端从wiki中解析")
    eps: int = Field(description="由旧服务端从wiki中解析，对于书籍条目为`话数`")

    total_episodes: int = Field(description="数据库中的章节数量")

    rating: Rating

    collection: Collection


class RelSubject(BaseModel):
    id: int
    type: int
    name: str
    name_cn: str
    images: Optional[Images]
    relation: str


class Episode(BaseModel):
    id: int
    type: int = Field(description="`0` 本篇，`1` SP，`2` OP，`3` ED")
    name: str
    name_cn: str
    sort: float = Field(description="同类条目的排序和集数")
    ep: float = Field(None, description="条目内的集数, 从`1`开始。非本篇剧集的此字段无意义")
    airdate: str
    comment: int
    duration: str
    desc: str = Field(description="简介")
    disc: int = Field(description="音乐曲目的碟片数")


class EpisodeDetail(Episode):
    subject_id: int

File: model/subject.go
// Copyright (c) 2022 Trim21 <trim21.me@gmail.com>
//
// SPDX-License-Identifier: AGPL-3.0-only
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, version 3.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>

package model

const subjectLocked = 2

type Subject struct {
	Image         string
	Summary       string
	Name          string
	Date          string // first release date
	NameCN        string
	Infobox       string
	CompatRawTags []byte // compat field for old tags
	OnHold        uint32
	Dropped       uint32
	Volumes       uint32
	Eps           uint32
	Wish          uint32
	Collect       uint32
	Doing         uint32
	ID            uint32
	PlatformID    uint16
	TypeID        SubjectType
	Ban           uint8
	Airtime       uint8 // air weekday, start from
	NSFW          bool
	Rating        Rating
	Redirect      uint32
}

func (s Subject) Locked() bool {
	return s.Ban == subjectLocked
}

type Count struct {
	Field1  uint32
	Field2  uint32
	Field3  uint32
	Field4  uint32
	Field5  uint32
	Field6  uint32
	Field7  uint32
	Field8  uint32
	Field9  uint32
	Field10 uint32
}

type Rating struct {
	Rank  int32
	Total uint32
	Count Count
	Score float64
}

type Platform struct {
	Alias        string `json:"alias"`
	Type         string `json:"type"`
	TypeCN       string `json:"type_cn"`
	WikiTpl      string `json:"wiki_tpl"`
	SearchString string `json:"search_string"`
	ID           int    `json:"id"`
	EnableHeader bool   `json:"enable_header"`
}

func (p Platform) String() string {
	if p.TypeCN != "" {
		return p.TypeCN
	}

	return p.Type
}

type Episode struct {
	Airdate     string
	Name        string
	NameCN      string
	Duration    string
	Description string
	Ep          float32
	SubjectID   uint32
	Sort        float32
	Comment     uint32
	ID          uint32
	Type        uint8
	Disc        uint8
}
